apps/rush/src/RushVersionManager.ts

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import * as path from 'path';
+import * as fsx from 'fs-extra';
+import * as semver from 'semver';
+
+import { Utilities } from '@microsoft/rush-lib';
+import * as rushLibCli from '@microsoft/rush-lib/lib/start';
+
+import RushWrapper from './RushWrapper';
+
+const RUSH_TRANSITIONAL_VERSION: string = '4.0.0';
+const MAX_INSTALL_ATTEMPTS: number = 5;
+
+export default class RushVersionManager {
+  private _rushDirectory: string;
+  private _currentPackageVersion: string;
+
+  constructor(homeDirectory: string, currentPackageVersion: string) {
+    this._rushDirectory = path.join(homeDirectory, '.rush');
+    this._currentPackageVersion = currentPackageVersion;
+  }
+
+  public ensureRushVersionInstalled(version: string): RushWrapper {
+    const isLegacyRushVersion: boolean = semver.lt(version, RUSH_TRANSITIONAL_VERSION);
+    const expectedRushPath: string = path.join(this._rushDirectory, `rush-${version}`);
+    const expectedRushInstalledFlagPath: string = path.join(expectedRushPath, 'last-install.flag');
+    if (!fsx.existsSync(expectedRushInstalledFlagPath)) {
+      // Need to install Rush
+      console.log(`Rush version ${version} is not currently installed. Installing...`);
```

Comment:
>console.log(`Rush version ${version} is not currently installed. Installing...`); [](start = 6, length = 81)

If two different CI jobs try to do this simultaneously, can something bad happen?  e.g. could we end up with a corrupted folder?

How did InstallRushOnlyIfNeeded.js handle this?